libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -80,29 +68,112 @@ bool KnowledgeBase::knownToBeZero(YulString _a)
 
 optional<u256> KnowledgeBase::valueIfKnownConstant(YulString _a)
 {
-	if (AssignedValue const* value = m_variableValues(_a))
-		if (Literal const* literal = get_if<Literal>(value->value))
-			return valueOfLiteral(*literal);
-	return {};
+	VariableOffset offset = explore(_a);
+	if (offset.reference == YulString{})
+		return offset.offset;
+	else
+		return nullopt;
+}
+
+optional<u256> KnowledgeBase::valueIfKnownConstant(Expression const& _expression)
+{
+	if (Identifier const* ident = get_if<Identifier>(&_expression))
+		return valueIfKnownConstant(ident->name);
+	else if (Literal const* lit = get_if<Literal>(&_expression))
+		return valueOfLiteral(*lit);
+	else
+		return {};
 }
 
-Expression KnowledgeBase::simplify(Expression _expression)
+KnowledgeBase::VariableOffset KnowledgeBase::explore(YulString _var)
 {
-	m_counter = 0;
-	return simplifyRecursively(std::move(_expression));
+	// Query the value mainly to potentially reset.
+	Expression const* value = valueOf(_var);
+	if (VariableOffset const* varOff = util::valueOrNullptr(m_offsets, _var))
+		return *varOff;
+
+	if (value)
+		if (optional<VariableOffset> offset = explore(*value))
+			return setOffset(_var, *offset);
+	return setOffset(_var, VariableOffset{_var, 0});
+
 }
 
-Expression KnowledgeBase::simplifyRecursively(Expression _expression)
+optional<KnowledgeBase::VariableOffset> KnowledgeBase::explore(Expression const& _value)
 {
-	if (m_counter++ > 100)
-		return _expression;
+	if (Literal const* literal = std::get_if<Literal>(&_value))
+		return VariableOffset{YulString{}, valueOfLiteral(*literal)};
+	else if (Identifier const* identifier = std::get_if<Identifier>(&_value))
+		return explore(identifier->name);
+	else if (FunctionCall const* f = get_if<FunctionCall>(&_value))
+		if (f->functionName.name == "add"_yulstring || f->functionName.name == "sub"_yulstring)
```

Comment:
This just assumes the semantics of "add" and "sub" without actually knowing the dialect. Is that fine?